experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts

```diff
@@ -0,0 +1,493 @@
+ * Allows for concurrent editing based on index, adjusting the locations of indexes as needed so they apply to the same logical place in the sequence when rebased and merged.
+ * @alpha
+ */
+export interface Sequence<TTypes extends AllowedTypes> extends UntypedField {
+	/**
+	 * Gets a node of this field by its index with unboxing.
+	 * Note that a node must exist at the given index.
+	 */
+	at(index: number): UnboxNodeUnion<TTypes>;
+
+	/**
+	 * Gets a boxed node of this field by its index.
+	 * Note that a node must exist at the given index.
+	 */
+	boxedAt(index: number): TypedNodeUnion<TTypes>;
+
+	/**
+	 * Calls the provided callback function on each child of this sequence, and returns an array that contains the results.
+	 * @param callbackfn - A function that accepts the child, its index, and this field.
+	 */
+	map<U>(callbackfn: (value: UnboxNodeUnion<TTypes>, index: number, array: this) => U): U[];
+
+	/**
+	 * Calls the provided callback function on each child of this sequence, and returns an array that contains the results.
+	 * @param callbackfn - A function that accepts the child, its index, and this field.
+	 */
+	mapBoxed<U>(callbackfn: (value: TypedNodeUnion<TTypes>, index: number, array: this) => U): U[];
+
+	readonly length: number;
+
+	// TODO: more and/or better editing APIs. As is, this can't express moves.
+	replaceRange(
+		index: number,
+		count: number,
+		content: Iterable<FlexibleNodeContent<TTypes>>,
+	): void;
+
+	[Symbol.iterator](): Iterator<TypedNodeUnion<TTypes>>;
+
+	// TODO: JS object traversal docs
+	readonly asArray: readonly UnboxNodeUnion<TTypes>[];
+}
+
+/**
+ * Field this stores exactly one child.
+ *
+ * @remarks
+ * Unboxes its content, so in schema aware APIs which do unboxing, the RequiredField itself will be skipped over and its content will be returned directly.
+ * @privateRemarks
+ * Finish renaming from ValueField to RequiredField
+ * @alpha
+ */
+export interface RequiredField<TTypes extends AllowedTypes> extends UntypedField {
+	readonly content: UnboxNodeUnion<TTypes>;
+	readonly boxedContent: TypedNodeUnion<TTypes>;
+	setContent(content: FlexibleNodeContent<TTypes>): void;
+}
+
+/**
+ * Field this stores zero or one child.
+ *
+ * @remarks
+ * Unboxes its content, so in schema aware APIs which do unboxing, the OptionalField itself will be skipped over and its content will be returned directly.
+ * @alpha
+ */
+export interface OptionalField<TTypes extends AllowedTypes> extends UntypedField {
+	readonly content?: UnboxNodeUnion<TTypes>;
+	readonly boxedContent?: TypedNodeUnion<TTypes>;
+	setContent(content: undefined | FlexibleNodeContent<TTypes>): void;
+}
+
+// #endregion
+
+// #region Typed
+
+/**
+ * Schema aware specialization of {@link UntypedField}.
+ * @alpha
+ */
+export type TypedField<TSchema extends FieldSchema> = TypedFieldInner<
+	TSchema["kind"],
+	TSchema["allowedTypes"]
+>;
+
+/**
+ * Helper for implementing {@link TypedField}.
+ * @alpha
+ */
+export type TypedFieldInner<
```

Comment:
Nit: is this abstraction necessary? Seems like it could be flattened into `TypedField` without being too messy, and then this wouldn't need to be in the API.

Reply:
There would be three copies of `TSchema["kind"]` and 3 of `TSchema["allowedTypes"]` which I think would hurt the readability a bit, especially since this list will be growing in the future (its missing a few uncommon kinds already like Forbidden and NodeKey, and will eventually likley get extra versions of the field kinds as their APIs evolve over time since they will get versioned for compat resons)